add blog post for in-tree cloud provider removal going beta (#44010)
kubernetes/website: content/en/blog/_posts/2023-12-20-disabling-in-tree-cloud-provider-goes-beta.md

Reviewer: BTW, localization teams will thank you if you wrap the Markdown at say 100 characters per line.

Author: wrapped at 80 chars

---
layout: blog

> title: "Removing In-Tree Cloud Providers Graduates to Beta Status"

Reviewer: ```suggestion
title: "Kubernetes 1.29: Cloud Provider Integrations Are Now Separate Components"
```

Author: update incoming

date: 2023-12-20

> slug: removing-in-tree-cloud-providers-beta-status

Reviewer: ```suggestion
slug: cloud-provider-integration-changs
```

---

> **Authors:** Michael McCune, Andrew Sy Kim

Reviewer: Usually we list authors' affiliations - see other articles for an example.

> 

Reviewer: Make sure to state the main point.
For example:
```suggestion
For Kubernetes v1.29, you need to use additional components to integrate your Kubernetes cluster
with a cloud infrastructure provider. By default, Kubernetes v1.29 does not run with the legacy,
compiled-in cloud provider integrations enabled.

```

Author: ii've added this to the beginning.

In 2018, the Kubernetes community agreed to form the Cloud Provider Special Interest Group (SIG), with a mission to externalize all cloud provider integrations and remove all the existing in-tree cloud provider integrations. In January 2019, the Kubernetes community approved the initial draft of [KEP-2395: Removing In-Tree Cloud Provider Code][kep2395]. This KEP defines a process by which we can remove cloud provider specific code from the core Kubernetes source tree. From the KEP:

> Motiviation behind this effort is to allow cloud providers to develop and make releases independent from the core Kubernetes release cycle. The de-coupling of cloud provider code allows for separation of concern between "Kubernetes core" and the cloud providers within the ecosystem. In addition, this ensures all cloud providers in the ecosystem are integrating with Kubernetes in a consistent and extendable way.

> After many years of development and collaboration across many contributors, SIG Cloud Provider is happy to announce that as of the Kubernetes 1.29 release, [KEP-2395][kep2395] will graduate to Beta status. The graduation to Beta is a big milestone as we’ve had to work through the complexities that come with migrating a core component in a distributed system like Kubernetes. With the graduation to Beta, all Kubernetes core components will have cloud provider integrations disabled by default. This brings us one step closer to its final removal which will happen when the KEP graduates to GA.

Reviewer: Whilst this is true, please consider writing this from the _reader_'s perspective. The typical reader cares about having to intervene to keep their cluster running much more than their level of interest in the organization layout (names of SIGs, etc).

Reviewer: I was thinking about the paragraph as a whole.

Things readers might not care about very much:
- the removal was led by SIG Cloud Provider
- there is a KEP
- the ID number of that KEP
- the feature stage (beta) for that KEP
- how complex the work has been

What they really want to know is: what action they need to take, how soon, and what happens if they don't act.

Author: i changed the paragraph to focus more on what you suggested.

> Although the new defaults will affect all Kubernetes users, they will require changes for users who are currently using the in-tree providers for Azure, GCE, and vSphere.

Reviewer: I'd expect to see something like:
```
These updated default settings affect a large proportion of Kubernetes users, and
**will require changes** for users who were previously using the in-tree provider integrations.
The legacy integrations offered compatibility with for Azure, AWS, GCE, OpenStack and vSphere; however,
for AWS and OpenStack DETAILS_GO_HERE.
```

I know there is a legacy in-tree AWS integration, so it's odd that we don't mention it here. Same for OpenStack.

(ie: don't make readers skip past this part of the page to understand what's going on)

Author: i specifically didn't mention the AWS or OpenStack providers here because they were removed prior to 1.27. but i will add them in to make it more clear.

Author: updated

> ## What is changing?

Reviewer: ```suggestion
## What has changed?
```

Author: done

Let's see what the [official Kubernetes documentation][fg] says about these feature gates:

> `DisableCloudProviders`: Disables any functionality in `kube-apiserver`, `kube-controller-manager` and `kubelet` related to the `--cloud-provider` component flag.

> `DisableKubeletCloudCredentialProviders` Disable the in-tree functionality in kubelet to authenticate to a cloud provider container registry for image pull credentials.

## What do you need to do?

If you are upgrading from Kubernetes 1.28+ and are not on Azure, GCE, or vSphere then you are in luck, there are no changes you will need to make. If you **are** on Azure, GCE, or vSphere, or you are upgrading from a version older than 1.28, then read on.

> Historically, Kubernetes has included code for a set of cloud providers that included AWS, Azure, GCE, OpenStack, and vSphere. Since the inception of [KEP-2395][kep2395] the community has been moving towards removal of that cloud provider code. The OpenStack provider code was removed in version 1.26, and the AWS provider code was removed in version 1.27. This means that users who are upgradging from one of the affected cloud providers and versions will need to modify their deployments.

Reviewer: IMO: this should appear before any mention of the term “SIG”.

Author: are you saying this paragraph should be moved to the intro section at the top?

```

_Please note that if you have other feature gate modifications on the command line, they will need to include these 2 feature gates._

### Upgrading on other providers

For providers other than Azure, GCE, or vSphere, good news, the external cloud controller manager should be already be in use. The code for AWS and OpenStack providers has been removed from Kubernetes before version 1.27 was released. Other providers beyond the 5 listed in this post were never included in Kubernetes and as such they began their life as external cloud controller managers.

### Upgrading from older Kubernetes versions

If you are upgrading from a Kubernetes release older than 1.26, and you are on AWS, Azure, GCE, OpenStack, or vSphere then you will need to enable the `--cloud-provider=external` flag, and follow the advice for installing and running a cloud controller manager for your provider.

Please read the documentation for [Cloud Controller Manager Administration][ccmadmin] and [Migrate Replicated Control Plane To Use Cloud Controller Manager][ccmha]. See below for links to specific cloud provider implementations.

## Where to find a cloud controller manager?

> At its core, this post is about the cloud controllers that were previously included in Kubernetes. As these components move out of the core Kubernetes code and into their own repositories it is important to note a few things.

Reviewer: ```suggestion
At its core, this announcement is about the cloud provider integrations that were previously included in Kubernetes. As these components move out of the core Kubernetes code and into their own repositories, it is important to note a few things:
```

Author: done

First, SIG Cloud Provider maintains a reference framework for developers who wish to create cloud controller managers for any provider. See the [cloud-provider repository][cloud-provider] for more information about how these controllers work and how to get started creating your own.

> Second, there are many cloud controller managers available for Kubernetes. This post is addressing the providers that have been historically included with Kubernetes but are now in the process of being removed. If you need a cloud controller manager for your provider and do not see it listed here, please reach out to the SIG Cloud Provider community (links below) for help and advice.

Reviewer: Can we frame it to help readers understand that, in future, the cloud provider integration plugin probably will be open source, but _might not be_?

> ### Cloud Controller Managers

Reviewer: ```suggestion
### Cloud provider integrations provided by the Kubernetes project {#cloud-provider-integrations}
```

Author: done

* OpenStack - https://github.com/kubernetes/cloud-provider-openstack
* vSphere - https://github.com/kubernetes/cloud-provider-vsphere

## Want to learn more?

Cloud providers and cloud controller managers serve a core function in Kubernetes. Cloud providers are often the substrate upon which Kubernetes is operated, and the cloud controller managers supply the essential lifeline between Kubernetes clusters and their physical infrastructure.

This post exposes one aspect of how the Kubernetes community interacts with the world of cloud infrastructure providers. If you are curious about this topic and want to learn more the Cloud Provider Special Interest Group (SIG) is the place to go. SIG Cloud Provider hosts bi-weekly meetings to discuss all manner of topics related to cloud providers and cloud controller managers in Kubernetes.

### SIG Cloud Provider

* Regular SIG Meeting: [Wednesdays at 9:00 PT (Pacific Time)](https://zoom.us/j/508079177?pwd=ZmEvMksxdTFTc0N1eXFLRm91QUlyUT09) (biweekly). [Convert to your timezone](http://www.thetimezoneconverter.com/?t=9:00&tz=PT%20%28Pacific%20Time%29).
* [Kubernetes slack][kslack] channel `#sig-cloud-provider`

> * [SIG Community page][sig]

Reviewer: Although this might feel like a feature, readers will experience this more as a loss: they didn't have to do a thing, and now they do. I would omit these details given we're basically imposing work on end users.

Author: just to be clear, remove the SIG details or the entire section on "Want to learn more" ?

---
layout: blog
title: "Kubernetes 1.29: Cloud Provider Integrations Are Now Separate Components"
date: 2023-12-20
slug: cloud-provider-integration-changes
---

**Authors:** Michael McCune (Red Hat), Andrew Sy Kim (Google)

For Kubernetes v1.29, you need to use additional components to integrate your
Kubernetes cluster with a cloud infrastructure provider. By default, Kubernetes

> v1.29 does not run with the legacy, compiled-in cloud provider integrations

Reviewer 2: I think we should reword this to:
```
By default, Kubernetes v1.29 disallows running the legacy compiled-in cloud provider integrations by default.
```

Author: i've updated the language a bit, but i did drop one of the "by default"s from that ;)

enabled.

In 2018, the Kubernetes community agreed to form the Cloud Provider Special

> Interest Group (SIG), with a mission to externalize all cloud provider

Reviewer 2: Do we still want to reference this older blog post somewhere? https://kubernetes.io/blog/2019/04/17/the-future-of-cloud-providers-in-kubernetes/

Author: i think that's nice background reading, but probably not essential to the new material. that said, i'd like to include a link, it might be easiest to just link in the text of this paragraph.

and **will require changes** for users who were previously using the in-tree
provider integrations. The legacy integrations offered compatibility with
Azure, AWS, GCE, OpenStack, and vSphere; however for AWS and OpenStack the
compiled-in integrations were removed in Kubernetes versions 1.26 and 1.27,
respectively.

## What has changed?

At the most basic level, two [feature gates][fg] are changing their default
value from false to true. Those feature gates, `DisableCloudProviders` and
`DisableKubeletCloudCredentialProviders`, control the way that the
[kube-apiserver][kapi], [kube-controller-manager][kcm], and [kubelet][kubelet]
invoke the cloud provider related code that is included in those components.
With these feature gates defaulting to true, the former behavior that was
provided through the `--cloud-provider` command line flag is changing to only
recognize the `external` value.

Let's see what the [official Kubernetes documentation][fg] says about these
feature gates:

> `DisableCloudProviders`: Disables any functionality in `kube-apiserver`,
> `kube-controller-manager` and `kubelet` related to the `--cloud-provider`
> component flag.

> > `DisableKubeletCloudCredentialProviders` Disable the in-tree functionality

Reviewer: ```suggestion
> `DisableKubeletCloudCredentialProviders`: Disable the in-tree functionality
```

Author: fixed

### Upgrading on Azure, GCE, or vSphere

There are two options for upgrading in this configuration: migrate to external
cloud controller managers, or continue using the in-tree provider code.
Although migrating to external cloud controller managers is recommended,
there are scenarios where continuing with the current behavior is desired.
Please choose the best option for your needs.

#### Migrate to external cloud controller managers

Migrating to use external cloud controller managers is the recommended upgrade
path as it aligns with the future of Kubernetes. To do this you will need to
enable the `--cloud-provider=external` command line flag for the
`kube-apiserver`, `kube-controller-manager`, and `kubelet` components. In
addition you will need to deploy a cloud controller manager for your provider.

Installing and running cloud controller managers is a larger topic than this
post can address, if you would like more information on this process please
read the documentation for [Cloud Controller Manager Administration][ccmadmin]
and [Migrate Replicated Control Plane To Use Cloud Controller Manager][ccmha].
See below for links to specific cloud provider implementations.

> #### Continue using in-tree provide code

Reviewer: ```suggestion
#### Continue using in-tree provider code
```

Author: fixed

listed commands:

```
--feature-gates=DisableCloudProviders=false,DisableKubeletCloudCredentialProviders=false
```

_Please note that if you have other feature gate modifications on the command
line, they will need to include these 2 feature gates._

### Upgrading on other providers

For providers other than Azure, GCE, or vSphere, good news, the external cloud
controller manager should be already be in use. The code for AWS and OpenStack
providers has been removed from Kubernetes before version 1.27 was released.
Other providers beyond the 5 listed in this post were never included in
Kubernetes and as such they began their life as external cloud controller
managers.

### Upgrading from older Kubernetes versions

If you are upgrading from a Kubernetes release older than 1.26, and you are on
AWS, Azure, GCE, OpenStack, or vSphere then you will need to enable the
`--cloud-provider=external` flag, and follow the advice for installing and
running a cloud controller manager for your provider.

Please read the documentation for
[Cloud Controller Manager Administration][ccmadmin] and
[Migrate Replicated Control Plane To Use Cloud Controller Manager][ccmha]. See

> below for links to specific cloud provider implementations.

Reviewer: (nit)
```suggestion
[below](#cloud-provider-integrations) for links to specific cloud provider implementations.
```

Author: added

these controllers work and how to get started creating your own.

Second, there are many cloud controller managers available for Kubernetes.
This post is addressing the providers that have been historically included
with Kubernetes but are now in the process of being removed. If you need a
cloud controller manager for your provider and do not see it listed here,
please reach out to the cloud provider you are integrating with or the
[Kubernetes SIG Cloud Provider community][sig] for help and advice. It is
worth noting that while most cloud controller managers are open source today,
this may not always be the case. Users should always contact their cloud
provider to learn if there are preferred solutions to utilize on their
infrastructure.

### Cloud provider integrations provided by the Kubernetes project {#cloud-provider-integrations}

* AWS - https://github.com/kubernetes/cloud-provider-aws
* Azure - https://github.com/kubernetes-sigs/cloud-provider-azure
* GCE - https://github.com/kubernetes/cloud-provider-gcp
* OpenStack - https://github.com/kubernetes/cloud-provider-openstack
* vSphere - https://github.com/kubernetes/cloud-provider-vsphere

## Want to learn more?

> Cloud providers and cloud controller managers serve a core function in

Reviewer: ```suggestion
Cloud provider integrations, and the core library for cloud controller managers, serve a core function in
```
?

Author: the cloud-provider framework isn't required per se, i was trying to call out the CCMs themselves here. like both the provider (the actual infrastructure) and the ccm are core to this functionality. maybe that was implying too much?

`kube-apiserver`, `kube-controller-manager`, and `kubelet` components. In
addition you will need to deploy a cloud controller manager for your provider.

Installing and running cloud controller managers is a larger topic than this
post can address, if you would like more information on this process please
read the documentation for [Cloud Controller Manager Administration][ccmadmin]
and [Migrate Replicated Control Plane To Use Cloud Controller Manager][ccmha].
See [below](#cloud-provider-integrations) for links to specific cloud provider
implementations.

#### Continue using in-tree provider code

If you wish to continue using Kubernetes with the in-tree cloud provider code,
you will need to modify the command line parameters for `kube-apiserver`,
`kube-controller-manager`, and `kubelet` to disable the feature gates for
`DisableCloudProviders` and `DisableKubeletCloudCredentialProviders`. To do
this, add the following command line flag to the arguments for the previously
listed commands:

```
--feature-gates=DisableCloudProviders=false,DisableKubeletCloudCredentialProviders=false
```

_Please note that if you have other feature gate modifications on the command

> line, they will need to include these 2 feature gates._

Reviewer 2: Should we add a note or warning here?:
```
WARNING: these feature gates will be locked to `true` in an upcoming release. Setting these feature gates to `false` should be a last resort and it is highly recommended to migrate to an external cloud controller manager.  
```

(or something like that)

Author: that sounds reasonable to me, i'll add it.

Author: updated

Interest Group (SIG)][oldblog], with a mission to externalize all cloud provider
integrations and remove all the existing in-tree cloud provider integrations.
In January 2019, the Kubernetes community approved the initial draft of
[KEP-2395: Removing In-Tree Cloud Provider Code][kep2395]. This KEP defines a
process by which we can remove cloud provider specific code from the core
Kubernetes source tree. From the KEP:

> Motiviation behind this effort is to allow cloud providers to develop and
> make releases independent from the core Kubernetes release cycle. The
> de-coupling of cloud provider code allows for separation of concern between
> "Kubernetes core" and the cloud providers within the ecosystem. In addition,
> this ensures all cloud providers in the ecosystem are integrating with
> Kubernetes in a consistent and extendable way.

After many years of development and collaboration across many contributors,
the default behavior for legacy cloud provider integrations is changing.
This means that users will need to confirm their Kubernetes configurations,
and in some cases run external cloud controller managers. These changes are

> taking affect in Kubernetes version 1.29, read on to learn if you are affected

Reviewer 3: ```suggestion
taking effect in Kubernetes version 1.29, read on to learn if you are affected
```

Author: done

[Kubernetes SIG Cloud Provider community][sig] for help and advice. It is
worth noting that while most cloud controller managers are open source today,
this may not always be the case. Users should always contact their cloud
provider to learn if there are preferred solutions to utilize on their
infrastructure.

### Cloud provider integrations provided by the Kubernetes project {#cloud-provider-integrations}

* AWS - https://github.com/kubernetes/cloud-provider-aws
* Azure - https://github.com/kubernetes-sigs/cloud-provider-azure
* GCE - https://github.com/kubernetes/cloud-provider-gcp
* OpenStack - https://github.com/kubernetes/cloud-provider-openstack
* vSphere - https://github.com/kubernetes/cloud-provider-vsphere

## Want to learn more?

Cloud providers and cloud controller managers serve a core function in
Kubernetes. Cloud providers are often the substrate upon which Kubernetes is
operated, and the cloud controller managers supply the essential lifeline
between Kubernetes clusters and their physical infrastructure.

This post exposes one aspect of how the Kubernetes community interacts with
the world of cloud infrastructure providers. If you are curious about this

> topic and want to learn more the Cloud Provider Special Interest Group (SIG)

Reviewer 3: ```suggestion
topic and want to learn more, the Cloud Provider Special Interest Group (SIG)
```

Author: done
